lib/app/admin/data-reset.js

```diff
@@ -0,0 +1,212 @@
+var q = require('q');
+var async = require('async');
+var _ = require('lodash');
+// default list of types of objects to remove, can be overriden when passed in request body
+// order of the list is important for building relations between them, parent items should be listed first
+var DEFAULT_ITEM_LIST = [
+  'messages',
+  'result',
+  'workflows',
+  'workorders',
+  'group',
+  'membership',
+  'file'
+];
+var DEFAULT_LIST_RELATIONS = {
+  workorders: 'workflows',
+  membership: 'group'
+};
+
+/**
+ * Rebuilds data in the solution - first it removes all items synchronously using mediator ':delete' topic,
+ * this is followed by ':create' topic using seed data located in ./data-demo
+ * @param {Object} mediator - mediator object used for communication
+ * @param {Array} removeList - list of element names e.g. DEFAULT_ITEM_LIST
+ * @returns {Promise} Promise, resolved if all items are successfully rebuilt.
+ */
+module.exports = function rebuildData(mediator, removeList) {
+  var deferred = q.defer();
+  var list = (removeList) ? removeList : DEFAULT_ITEM_LIST;
+  deleteData(mediator, list)
+    .then(function() {
+      buildData(mediator, list);
+    })
+    .then(function() {
+      deferred.resolve();
+    })
+    .catch(function(err) {
+      deferred.reject(err);
+    });
+  return deferred.promise;
+};
+/**
+ * Deletes all data in the solution
+ * @param {Object} mediator - mediator object used for communication
+ * @param {Array} removeList - list of element names e.g. DEFAULT_ITEM_LIST
+ * @returns {Promise} Promise, resolved if all data are successfully deleted.
+ */
+function deleteData(mediator, list) {
+
+
+  var deferred = q.defer();
+
+  async.eachSeries(list, function(item, eachSeriesCb) {
+    // delete all items of a type
+    deleteAllItems(mediator, item)
+      .then(function() {
+        eachSeriesCb();
+      }).catch(function(err) {
+        deferred.reject(err);
```

Comment:
Within the `eachSeries` function, it should not call any other flow control function other than the `eachSeriesCb` function.

So if there is an error, it should call `eachSeriesCb` with the error. 

The async.eachSeries function will then call the function to end the series.

If you use the `deferred.reject` here, it can cause a memory leak as the end function of the `async.eachSeries` is never called.